src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketTagClientImpl.java

```diff
@@ -61,24 +60,35 @@ public Stream<BitbucketTag> getRemoteTags() {
                 .flatMap(Collection::stream);
     }
 
-    public BitbucketTag getRemoteTag(String tagName) {
-        HttpUrl.Builder urlBuilder = null;
-        try {
-            urlBuilder = bitbucketRequestExecutor.getCoreRestPath().newBuilder()
+    @Override
+    public BitbucketTag getRemoteTag(String tagName) throws NotFoundException {
+        HttpUrl url = bitbucketRequestExecutor.getCoreRestPath().newBuilder()
                     .addPathSegment("projects")
                     .addPathSegment(projectKey)
                     .addPathSegment("repos")
                     .addPathSegment(repositorySlug)
-                    .addPathSegment("tags")
-                    .addPathSegment(URLEncoder.encode(tagName, "UTF-8"));
-        } catch (UnsupportedEncodingException e) {
-            throw new RuntimeException(e);
+                    .addPathSegment("commits")
+                    .addQueryParameter("until", tagName)
+                    .addQueryParameter("start", "0")
+                    .addQueryParameter("limit", "1")
+                    .build();
+
+        BitbucketPage<BitbucketCommit> firstPage =
+                bitbucketRequestExecutor.makeGetRequest(url,
+                        new TypeReference<BitbucketPage<BitbucketCommit>>() {
+                        }).getBody();
+
+        BitbucketCommit commit = BitbucketPageStreamUtil.toStream(firstPage, new BitbucketTagClientImpl.OnlyPageFetcherImpl())
+                .map(BitbucketPage::getValues)
+                .flatMap(Collection::stream)
+                .findFirst().orElse(null);
+
+        if (commit == null) {
+            throw new NotFoundException(String.format("Unable to locate tag with name %s", tagName), firstPage.toString());
         }
```

Comment:
@mhenschke-atl @dkjellin, thoughts here?
a tag that hasnt been found will throw a 404 prompting the scmSource to catch the error, therefore we never get here. I dont think theres a case where you can get a valid commit, with no commit details. Should i just remove this check & assume that if we make it here we have a correct value? 

It also doesnt hurt to just check this.

Reply:
I haven't tried it, but it looks like the endpoint only throws a 404 if [the repo doesn't exist](https://developer.atlassian.com/server/bitbucket/rest/v903/api-group-repository/#api-api-latest-projects-projectkey-repos-repositoryslug-commits-get). Have you tried this to confirm?
This is exceptional behaviour for the code path (probably a race condition between executor picking up a job and a tag being deleted on repo), but I notice `BitbucketSCMSource#fetchBitbucketTag` is returning an optional anyway, so maybe this method could too? Then you can just flatmap it and handle it as though the scm helper is missing- that returns a null revision which should causes the job to fail elegantly?

Reply:
Yep, have tried with a ref that does exist. (e.g. intercept the call and change it to `tag1` or something that doesnt exist). 

If we make this method return an optional then we a returning an optional<optional<bitbucketTag>> but i think i can cut that out and send it back. Will try it out

Reply:
You mean in the fetch? Yes you can `flatmap` to solve that issue (or just return null here as in your commit client- either works)